src/layout/Popover.js

```diff
@@ -1,36 +1,230 @@
-              fontWeight: 'normal',
-              textAlign: 'left',
-              background: 'white',
-              border: '1px solid #ebebeb',
-              ...style,
-            }}
-          >
-            {children}
-          </div>
-        </div>
-      </OutsideClickHandler>
+import {
+  moveX,
+  moveY,
+  middleX,
+  middleY,
+  area,
+  intersection,
+  fromObject,
+} from '../utils/rectangle'
+import { addEventListener, setNodeStyles } from '../utils/dom'
+
+let { min, max, floor } = Math
+
+let getPopoverTopleft = ({ windowRect, anchorRect, popoverRect, position }) => {
+  /*
+   * These rectangles represent the places we can render a popover, relative to
+   * its anchor element. They are positioned by applying translation functions:
+   * - The first two translations move a rectangle so that one of its boundaries
+   *   coincides with one of the boundaries of the anchor element.
+   * - The next two move a rectangle such that its visible area is maximized,
+   *   prioritizing the visibility of its topleft area.
+   *
+   * NOTE: The -1 in one of the translation functions is a hack to avoid text
+   * wrapping from making the popover content overflow into the anchor element.
+   */
+  let positioned = F.mapIndexed(
+    (translate, position) => ({ position, rectangle: translate(popoverRect) }),
+    {
+      top: _.flow(
+        pop => moveY(anchorRect.top - pop.bottom)(pop),
+        pop => moveX(middleX(anchorRect) - middleX(pop))(pop),
+        pop => moveX(min(0, windowRect.right - pop.right - 1))(pop),
+        pop => moveX(max(0, windowRect.left - pop.left))(pop)
+      ),
+      right: _.flow(
+        pop => moveX(anchorRect.right - pop.left)(pop),
+        pop => moveY(middleY(anchorRect) - middleY(pop))(pop),
+        pop => moveY(min(0, windowRect.bottom - pop.bottom))(pop),
+        pop => moveY(max(0, windowRect.top - pop.top))(pop)
+      ),
+      bottom: _.flow(
+        pop => moveY(anchorRect.bottom - pop.top)(pop),
+        pop => moveX(middleX(anchorRect) - middleX(pop))(pop),
+        pop => moveX(min(0, windowRect.right - pop.right - 1))(pop),
+        pop => moveX(max(0, windowRect.left - pop.left))(pop)
+      ),
+      left: _.flow(
+        pop => moveX(anchorRect.left - pop.right)(pop),
+        pop => moveY(middleY(anchorRect) - middleY(pop))(pop),
+        pop => moveY(min(0, windowRect.bottom - pop.bottom))(pop),
+        pop => moveY(max(0, windowRect.top - pop.top))(pop)
+      ),
+    }
+  )
+
+  // A bounding rectangle is one determined by the anchor element and the
+  // viewport borders.
+  let boundingRect = position =>
+    _.set(
+      { top: 'bottom', right: 'left', bottom: 'top', left: 'right' }[position],
+      anchorRect[position],
+      windowRect
+    )
+
+  // These heuristics will be used to choose among the possible popovers that
+  // can be rendered. The order matters: earlier functions have higher weight
+  // than latter ones.
+  let heuristics = [
```

Comment:
This is necessary. The first condition should be to prefer the popover with the largest visible area. We don't want clipped popovers. But imagine they all have the same area because there's enough space. Then we'd choose the one the user has specified. However, if they haven't specified a position, or if the position the user requested ends up with the popover being clipped, we've still gotta choose from a possible non-empty set of popovers, which is what the last function is for.
You're welcome to try and improve the heuristics, but be aware it was already quite hard to bring it down to three without doing lots of conditional for edge cases.